### PR TITLE
Fixes #16682 - Compress webpack compiled assets

### PR DESCRIPTION
**Pull request by tbrisker**

### CODE REVIEW

**Comment by gailsteiger** [comment]
What does this do?

**Comment by ohadlevy** [comment]
@tbrisker when I tested it on my prod system I've  noticed the following error:

```
RAILS_ENV=production rake assets:precompile webpack:compile
DEPRECATION WARNING: You are using the a deprecated processor interface #<Proc:0x00000006c203e8@/home/foreman/gems/gems/autoprefixer-rails-5.2.1.3/lib/autoprefixer-rails/sprockets.rb:25>.
Please update your processor interface:
https://github.com/rails/sprockets/blob/master/guides/extending_sprockets.md#supporting-all-versions-of-sprockets-in-processors
 (called from install at /home/foreman/gems/gems/autoprefixer-rails-5.2.1.3/lib/autoprefixer-rails/sprockets.rb:25)
/home/foreman/git/foreman/node_modules/.bin/webpack --config /home/foreman/git/foreman/config/webpack.config.js --bail
TypeError: object is not a function
    at Gzip.onError (zlib.js:161:5)
    at Gzip.emit (events.js:95:17)
    at validChunk (_stream_writable.js:154:12)
    at Gzip.Writable.write (_stream_writable.js:182:12)
    at Gzip.Writable.end (_stream_writable.js:341:10)
    at zlibBuffer (zlib.js:146:10)
    at Object.exports.gzip (zlib.js:116:3)
    at CompressionPlugin.<anonymous> (/home/foreman/git/foreman/node_modules/compression-webpack-plugin/index.js:68:10)
    at /home/foreman/git/foreman/node_modules/compression-webpack-plugin/node_modules/async/lib/async.js:111:13
    at Array.forEach (native)
    at _each (/home/foreman/git/foreman/node_modules/compression-webpack-plugin/node_modules/async/lib/async.js:32:24)
    at Object.async.each (/home/foreman/git/foreman/node_modules/compression-webpack-plugin/node_modules/async/lib/async.js:110:9)
    at CompressionPlugin.<anonymous> (/home/foreman/git/foreman/node_modules/compression-webpack-plugin/index.js:55:10)
    at Compilation.applyPluginsAsync (/home/foreman/git/foreman/node_modules/tapable/lib/Tapable.js:71:13)
    at Compilation.<anonymous> (/home/foreman/git/foreman/node_modules/webpack/lib/Compilation.js:572:10)
    at Compilation.next (/home/foreman/git/foreman/node_modules/tapable/lib/Tapable.js:67:11)
    at Compilation.<anonymous> (/home/foreman/git/foreman/node_modules/webpack/lib/optimize/UglifyJsPlugin.js:140:4)
    at Compilation.applyPluginsAsync (/home/foreman/git/foreman/node_modules/tapable/lib/Tapable.js:71:13)
    at Compilation.<anonymous> (/home/foreman/git/foreman/node_modules/webpack/lib/Compilation.js:567:9)
    at Compilation.next (/home/foreman/git/foreman/node_modules/tapable/lib/Tapable.js:67:11)
    at ExtractTextPlugin.<anonymous> (/home/foreman/git/foreman/node_modules/extract-text-webpack-plugin/index.js:309:4)
    at Compilation.applyPluginsAsync (/home/foreman/git/foreman/node_modules/tapable/lib/Tapable.js:71:13)
    at Compilation.<anonymous> (/home/foreman/git/foreman/node_modules/webpack/lib/Compilation.js:563:8)
    at Compilation.next (/home/foreman/git/foreman/node_modules/tapable/lib/Tapable.js:67:11)
    at ExtractTextPlugin.<anonymous> (/home/foreman/git/foreman/node_modules/extract-text-webpack-plugin/index.js:285:5)
    at /home/foreman/git/foreman/node_modules/async/lib/async.js:52:16
    at Object.async.forEachOf.async.eachOf (/home/foreman/git/foreman/node_modules/async/lib/async.js:236:30)
    at Object.async.forEach.async.each (/home/foreman/git/foreman/node_modules/async/lib/async.js:209:22)
    at ExtractTextPlugin.<anonymous> (/home/foreman/git/foreman/node_modules/extract-text-webpack-plugin/index.js:237:10)
    at Compilation.applyPluginsAsync (/home/foreman/git/foreman/node_modules/tapable/lib/Tapable.js:71:13)
```

**Comment by ohadlevy** [comment]
Also, am I right to assume this generates only the compress assets? I think we need both of them for fall back purpose.

**Comment by tbrisker** [comment]
@gailsteiger @ohadlevy This should create both compressed and uncompressed assets, letting apache choose which one to serve depending on the browser support.
I also see the same errors in the jenkins test, it seems they broke something in their own api that makes it not work on certain node versions, will dive into it later to find a workaround.

**Comment by tbrisker** [comment]
test failures are unrelated, see http://projects.theforeman.org/issues/16700

**Comment by dLobatog** [comment]
Isn't apache mod_deflate doing this already in production? I'm fine with having this & https://github.com/theforeman/foreman/pull/3295/files for development, but IDK if it'll cause any issues in production to compress things twice

**Comment by ohadlevy** [comment]
I know the assets pipeline does it too, and i remember @lzap introducing it initially. any idea @lzap ?

**Comment by lzap** [comment]
I was only adding apache configuration bit that servers .gz files when present with correct MIME headers, so Apache does not need to deflate them over and over again. Here is the bit, its somewhere in our puppet codebase:

``` conf
<Directory /usr/share/foreman/public/assets>

  # Use standard http expire header for assets instead of ETag
  <IfModule mod_expires.c>
    Header unset ETag
    FileETag None
    ExpiresActive On
    ExpiresDefault "access plus 1 year"
  </IfModule>

  # Return compressed assets if they are precompiled
  <IfModule mod_rewrite.c>
    RewriteEngine on
    # Make sure the browser supports gzip encoding and file with .gz added
    # does exist on disc before we rewrite with the extension
    RewriteCond %{HTTP:Accept-Encoding} \b(x-)?gzip\b
    RewriteCond %{REQUEST_FILENAME}.gz -s
    RewriteRule ^(.+) $1.gz [L]
    # Set headers for all possible assets which are compressed
    <FilesMatch \.css\.gz$>
      ForceType text/css
      Header set Content-Encoding gzip
      SetEnv no-gzip
    </FilesMatch>
    <FilesMatch \.js\.gz$>
      ForceType text/javascript
      Header set Content-Encoding gzip
      SetEnv no-gzip
    </FilesMatch>
    <FilesMatch \.svg\.gz$>
      ForceType image/svg+xml
      Header set Content-Encoding gzip
      SetEnv no-gzip
    </FilesMatch>
  </IfModule>

</Directory>
```

**Comment by tbrisker** [comment]
After a chat with @lzap seems that our other assets are compressed by the assets pipeline and the apache bit only has to do with knowing to serve them, so I believe this is needed.

**Comment by lzap** [comment]
How do one use this new code? Is the rake command the same?

**Comment by tbrisker** [comment]
@lzap it is run when executing webpack in production automatically, which is what we do in the packaging: https://github.com/theforeman/foreman-packaging/blob/rpm/develop/foreman/foreman.spec#L675

**Comment by tbrisker** [comment]
Also removed workaround for node 0.10 since we no longer build on it afaik.

File: package.json [hunk 1]
```diff
@@ -16,6 +16,7 @@
     "babel-preset-es2015": "~6.6.0",
     "babel-preset-react": "^6.5.0",
     "babel-register": "^6.9.0",
+    "compression-webpack-plugin": "~0.3.1",
```
Comment (lzap) [inline package.json:19]: Does this need new RPM package dependency?

Comment (tbrisker) [reply package.json:19]: yes, I opened https://github.com/theforeman/foreman-packaging/pull/1453 for that

**Comment by dLobatog** [comment]
Thanks @tbrisker @lzap  @ohadlevy @gailsteiger !

**Pull request merged**